### PR TITLE
MobileNetV2

### PR DESCRIPTION
**Pull request by dwSun**
## Description ##
MobileNetV2 model from the
    `"Inverted Residuals and Linear Bottlenecks: Mobile Networks for  Classification, Detection and Segmentation"
    <https://arxiv.org/abs/1801.04381>`_ paper.

### Essentials ###
- [x] Passed code style checking (`pylint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- [x] Code is well-documented: 
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
- Not sure this is the correct implement, but this model did works.

### CODE REVIEW

**Comment by dwSun** [comment]
I have put the MobileNetV2 class in the gluon vision model zoo.

I don't have pretrained models, so this should be modified:
> We provide pre-trained models for all the listed models.
These models can constructed by passing ``pretrained=True``:

BTW, pylint is so strict that I have to disable some rules.

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py [hunk 1]
```diff
@@ -158,3 +166,158 @@ def mobilenet0_25(**kwargs):
...
+            self.flatten = nn.Flatten(prefix='flat_')
+
+    def hybrid_forward(self, F, x):
+        x = self.features(x)
+        x = self.output(x)
+        x = self.flatten(x)
+        return x
+
+
+def get_mobilenetv2(w, pretrained=False, ctx=cpu(),
```
Comment (szha) [inline python/mxnet/gluon/model_zoo/vision/mobilenet.py]: add a version argument to get_mobilenet.

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py [hunk 2]
```diff
@@ -158,3 +166,158 @@ def mobilenet0_25(**kwargs):
...
+    if pretrained:
+        from ..model_store import get_model_file
+        version_suffix = '{0:.2f}'.format(w)
+        if version_suffix in ('1.00', '0.50'):
+            version_suffix = version_suffix[:-1]
+        net.load_params(get_model_file('mobilenetv2_%s' % version_suffix, root=root), ctx=ctx)
+    return net
+
+
+def mobilenetv2_1_0(**kwargs):
```
Comment (szha) [inline python/mxnet/gluon/model_zoo/vision/mobilenet.py]: provide the same variants as the v1 mobilenet

File: example/image-classification/symbols/mobilenetv2.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# -*- coding:utf-8 -*-
+# pylint: disable= arguments-differ,line-too-long,invalid-name
```
Comment (szha) [inline example/image-classification/symbols/mobilenetv2.py]: Don't use file-level lint disable. Please use inline lint disables to identify the lines that violate the lint rules and we can try to decide whether to fix or disable.

File: example/image-classification/symbols/mobilenetv2.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+'''
+MobileNetV2, implemented in Gluon.
```
Comment (szha) [inline example/image-classification/symbols/mobilenetv2.py]: delete duplicate and load from gluon model zoo instead.

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py [hunk 5]
```diff
@@ -17,27 +17,76 @@
...
+__modify__ = 'dwSun'
+__modified_date__ = '18/01/31'
 
 import os
 
+from ... import nn
 from ....context import cpu
 from ...block import HybridBlock
-from ... import nn
+
```
Comment (dwSun) [inline python/mxnet/gluon/model_zoo/vision/mobilenet.py]: iosrt made this change.

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py [hunk 6]
```diff
@@ -17,27 +17,76 @@
...
 
 # Helpers
+# pylint: disable= too-many-arguments
```
Comment (dwSun) [inline python/mxnet/gluon/model_zoo/vision/mobilenet.py]: This is mobilenet v1 code, add this to pass the style check. Code not changed.

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py [hunk 7]
```diff
@@ -53,15 +102,18 @@ class MobileNet(HybridBlock):
...
             with self.features.name_scope():
-                _add_conv(self.features, channels=int(32*multiplier), kernel=3, pad=1, stride=2)
-                dw_channels = [int(x*multiplier) for x in [32, 64]+[128]*2+[256]*2+[512]*6+[1024]]
-                channels = [int(x*multiplier) for x in [64]+[128]*2+[256]*2+[512]*6+[1024]*2]
+                _add_conv(self.features, channels=int(32 * multiplier), kernel=3, pad=1, stride=2)
+                # pylint: disable= line-too-long
+                dw_channels = [int(x * multiplier) for x in [32, 64] + [128] * 2 + [256] * 2 + [512] * 6 + [1024]]
+                channels = [int(x * multiplier) for x in [64] + [128] * 2 + [256] * 2 + [512] * 6 + [1024] * 2]
                 strides = [1, 2] * 3 + [1] * 5 + [2, 1]
+                # pylint: disable= invalid-name
```
Comment (dwSun) [inline python/mxnet/gluon/model_zoo/vision/mobilenet.py]: This is mobilenet v1 code, add this to pass the style check.
autopep8 made those changes.

**Comment by dwSun** [comment]
Requested changes have been made. I am trying to make my code more clear by merging commit logs. 
Seems the situation not going as I wish.

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py (hunk 7 above)
Comment (szha) [reply python/mxnet/gluon/model_zoo/vision/mobilenet.py]: please use `make pylint` as the standard for checking code style. we have a 100 char-per-line limit.

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py (hunk 6 above)
Comment (szha) [reply python/mxnet/gluon/model_zoo/vision/mobilenet.py]: please remove and use our pylintrc as standard.

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py [hunk 8]
```diff
@@ -74,13 +126,102 @@ def hybrid_forward(self, F, x):
...
+
+                c_bgn = int(96 * multiplier)
+                c_end = int(160 * multiplier)
+                self.features.add(BottleNeck(c_in=c_bgn, c_out=c_end, t=6, s=2))
+                self.features.add(BottleNeck(c_in=c_end, c_out=c_end, t=6, s=1))
+                self.features.add(BottleNeck(c_in=c_end, c_out=c_end, t=6, s=1))
+
+                c_bgn = int(160 * multiplier)
+                c_end = int(320 * multiplier)
+                self.features.add(BottleNeck(c_in=c_bgn, c_out=c_end, t=6, s=1))
```
Comment (szha) [inline python/mxnet/gluon/model_zoo/vision/mobilenet.py]: the construction looks like it could be refactored. can you use similar implementation from mobilenet v1?

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py [hunk 9]
```diff
@@ -89,69 +230,111 @@ def get_mobilenet(multiplier, pretrained=False, ctx=cpu(),
...
     <https://arxiv.org/abs/1704.04861>`_ paper, with width multiplier 1.0.
 
+    MobileNetV2 model from the
+    `"Inverted Residuals and Linear Bottlenecks:
+      Mobile Networks for Classification, Detection and Segmentation"
+    <https://arxiv.org/abs/1801.04381>`_ paper.
+
     Parameters
     ----------
+    ver : int, default 1
```
Comment (szha) [inline python/mxnet/gluon/model_zoo/vision/mobilenet.py]: version. same below

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py [hunk 10]
```diff
@@ -147,57 +143,21 @@ def __init__(self, multiplier=1.0, classes=1000, **kwargs):
...
+
+                c_bgns = [int(x * multiplier) for x in [32] + [16] + [24] * 2
+                          + [32] * 3 + [64] * 4 + [96] * 3 + [160] * 3]
+                c_ends = [int(x * multiplier) for x in [16] + [24] * 2 + [32] * 3
+                          + [64] * 4 + [96] * 3 + [160] * 3 + [320]]
+                ts = [1] + [6] * 16
+                strides = [1, 2] * 2 + [1] * 2 + [2] + [1] * 3 + [1] * 3 + [2] + [1] * 3
+
+                for c_bgn, c_end, t, s in zip(c_bgns, c_ends, ts, strides):
+                    self.features.add(BottleNeck(c_in=c_bgn, c_out=c_end, t=t, s=s))
```
Comment (dwSun) [inline python/mxnet/gluon/model_zoo/vision/mobilenet.py]: The code here looks ugly and very hard to follow the data flow.
Is this kind of code really necessary?

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py [hunk 11]
```diff
@@ -74,13 +122,66 @@ def hybrid_forward(self, F, x):
...
+
+                last_channels = int(1280 * multiplier) if multiplier > 1.0 else 1280
+                _add_conv(self.features, last_channels)
+
+                self.features.add(nn.GlobalAvgPool2D())
+
+            self.output = nn.Conv2D(classes, 1, use_bias=False, prefix='pred_')
+            self.flatten = nn.Flatten(prefix='flat_')
```
Comment (szha) [inline python/mxnet/gluon/model_zoo/vision/mobilenet.py]: don't introduce new field. flatten should be included in output.

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py (hunk 6 above)
Comment (szha) [reply python/mxnet/gluon/model_zoo/vision/mobilenet.py]: ping

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py [hunk 12]
```diff
@@ -17,27 +17,72 @@
...
+    Parameters
+    ----------
+    c_in : int
+        Number of input channels.
+    c_out : int
+        Number of output channels.
+    t : int
+        Layer expansion ratio.
+    s : int
+        strides
```
Comment (szha) [inline python/mxnet/gluon/model_zoo/vision/mobilenet.py]: Use the argument naming from https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/model_zoo/vision/resnet.py#L95-L102

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py (hunk 10 above)
Comment (szha) [reply python/mxnet/gluon/model_zoo/vision/mobilenet.py]: 20 lines of `self.features.add(BottleNeck...` is unnecessarily verbose. I personally prefer the more compact code.

Could you use the same variable naming as in V1?

Comment (dwSun) [reply python/mxnet/gluon/model_zoo/vision/mobilenet.py]: I prefer compact code too. But shouldn't we keep our code simply and clear.
Some people use the word 'scary' to describe code inside caffe and torch. Maybe it will be helpful to keep the code simple and stupid.
This kind of code take twice the time to write and more time to understand. As a programmer,  I am a bit lazy. So, personally, I prefer previous version.

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py (hunk 12 above)
Comment (szha) [reply python/mxnet/gluon/model_zoo/vision/mobilenet.py]: s -> strides

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py [hunk 13]
```diff
@@ -74,13 +123,68 @@ def hybrid_forward(self, F, x):
...
+                for in_c, c, t, s in zip(in_channels_group, channels_group, ts, strides):
+                    self.features.add(BottleNeck(in_channels=in_c, channels=c, t=t, stride=s))
+
+                last_channels = int(1280 * multiplier) if multiplier > 1.0 else 1280
+                _add_conv(self.features, last_channels)
+
+                self.features.add(nn.GlobalAvgPool2D())
+
+            self.output = nn.HybridSequential(prefix='output_')
+            self.output.add(
```
Comment (szha) [inline python/mxnet/gluon/model_zoo/vision/mobilenet.py]: `with self.output.name_scope():`
and manual prefix can be skipped.

Comment (dwSun) [reply python/mxnet/gluon/model_zoo/vision/mobilenet.py]: A name_scope is indeed necessary.

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py [hunk 14]
```diff
@@ -74,13 +123,69 @@ def hybrid_forward(self, F, x):
...
+            self.features = nn.HybridSequential(prefix='features_')
+            with self.features.name_scope():
+                _add_conv(self.features, int(32 * multiplier), kernel=3, stride=2, pad=1)
+
+                in_channels_group = [int(x * multiplier) for x in [32] + [16] + [24] * 2
+                                     + [32] * 3 + [64] * 4 + [96] * 3 + [160] * 3]
+                channels_group = [int(x * multiplier) for x in [16] + [24] * 2 + [32] * 3
+                                  + [64] * 4 + [96] * 3 + [160] * 3 + [320]]
+                ts = [1] + [6] * 16
+                strides = [1, 2] * 2 + [1] * 2 + [2] + [1] * 3 + [1] * 3 + [2] + [1] * 3
```
Comment (zhreshold) [inline python/mxnet/gluon/model_zoo/vision/mobilenet.py]: I think there's a mistake in paper:
screen shot 2018-02-13 at 5 54 45 pm
Therefore the strides here should be:
```
strides = [1, 2] * 2 + [1] * 3 + [2] + [1] * 2 + [1] * 3 + [2] + [1] * 3
```
More inputs and eyes are welcome if you find this not correct @dwSun @szha

Comment (zhreshold) [reply python/mxnet/gluon/model_zoo/vision/mobilenet.py]: BTW, I am comparing http://ethereon.github.io/netscope/#/gist/d01b5b8783b4582a42fe07bd46243986
with 
[plot.gv.pdf](https://github.com/apache/incubator-mxnet/files/1722162/plot.gv.pdf)

Comment (dwSun) [reply python/mxnet/gluon/model_zoo/vision/mobilenet.py]: The strides are correct in paper. We are applying stride 2 to last block of previous group of blocks.

Comment (zhreshold) [reply python/mxnet/gluon/model_zoo/vision/mobilenet.py]: @dwSun from 28^2x32 to 28^2x64, it should be stride 1. I am observing 14^2x64 feature map, which corresponds to stride 2. 
The input sizes and strides in this table is conflicting, but I am not sure which one is correct unless we compute the Mult-add and compare with the claims in paper.

Comment (dwSun) [reply python/mxnet/gluon/model_zoo/vision/mobilenet.py]: My mistake. Those 2 lines indeed can't make sense. Will correct it in next commit.

**Comment by dwSun** [comment]
for reference, this is the struct I am using:
```
input shape     stride  output shape
# conv2d
224x224x3       2       112x112x32

# bottleneck
112x112x32      1       112x112x16

# bottleneck
112x112x16      2       56x56x24
56x56x24        1       56x56x24

# bottleneck
56x56x24        2       28x28x32
28x28x32        1       28x28x32
28x28x32        1       28x28x32

# bottleneck
28x28x32        1       28x28x64
28x28x64        1       28x28x64
28x28x64        1       28x28x64
28x28x64        1       28x28x64

# bottleneck
28x28x64        2       14x14x96
14x14x96        1       14x14x96
14x14x96        1       14x14x96

# bottleneck
14x14x96        2       7x7x160
7x7x160         1       7x7x160
7x7x160         1       7x7x160

# bottleneck
7x7x160         1       7x7x320

# conv2d
7x7x320         1       7x7x1280

# avgpool
7x7x1280        _       1x1x1280

# the output
1x1x1280        _       1x1xk

```

**Comment by szha** [comment]
Would you mind adding the variants to `tests/python/unittest/test_gluon_model_zoo.py`?

**Comment by szha** [comment]
@zhreshold please approve once you confirm that it's implemented correctly.

**Comment by dwSun** [comment]
unittest finished.

File: python/mxnet/gluon/model_zoo/vision/mobilenet.py [hunk 15]
```diff
@@ -74,13 +124,71 @@ def hybrid_forward(self, F, x):
...
+        Number of classes for the output layer.
+    """
+
+    def __init__(self, multiplier=1.0, classes=1000, **kwargs):
+        super(MobileNetV2, self).__init__(**kwargs)
+        with self.name_scope():
+            self.features = nn.HybridSequential(prefix='features_')
+            with self.features.name_scope():
+                _add_conv(self.features, int(32 * multiplier), kernel=3,
+                          stride=2, pad=1, active=False)
```
Comment (zhreshold) [inline python/mxnet/gluon/model_zoo/vision/mobilenet.py]: no `relu` for the first conv?

Comment (dwSun) [reply python/mxnet/gluon/model_zoo/vision/mobilenet.py]: The MobileNetV2 paper not mentioned relu after first conv2d.
Also the channels of first conv2d is 32, while the channels of first bottleneck is 16, this doesn't sense.
I have tested 3 different model architectures:
- with relu in first conv2d
- without relu in first conv2d
- without relu and use 16 instead of 32 as output channels in first conv2d

The result is a bit confusing, since performance among those architectures looks similar.
Here is the result:
test

The dataset contains 100 classes, 19W pics, all gray-scale image. This dataset maybe too small. But I don't have enough resource to test on coco or imagenet.

**Comment by dwSun** [comment]
Add relu in first conv2d as requested.

**Comment by Jing-Luo** [comment]
Any plan to release the pretrained model? I try to train the model by my own and it looks not so well...

For more details, please look at

https://discuss.gluon.ai/t/topic/5295

**Comment by szha** [comment]
@Jing-Luo It's currently WIP.
